src/hypothesis/strategies.py

```diff
@@ -1589,6 +1577,94 @@ def label(self):
     return accept
 
 
+@defines_strategy_with_reusable_values
+@cacheable
+@composite
+def complex_numbers(
+        draw,
+        min_magnitude=0,
+        max_magnitude=None,
+        allow_infinity=None,
+        allow_nan=None
+):
+    """Returns a strategy that generates complex numbers.
+
+    This strategy draws complex numbers with constrained magnitudes.
+    The ``min_magnitude`` and ``max_magnitude`` parameters should be
+    non-negative :class:`~python:numbers.Real` numbers; values
+    of ``None`` correspond to zero and infinite values respectively.
+
+    If ``min_magnitude`` is positive or ``max_magnitude`` is finite, it
+    is an error to enable ``allow_nan``.
+
+    If ``max_magnitude`` is finite, it is an error to enable
+    ``allow_infinity``.
+
+    Examples from this strategy shrink by shrinking their real and
+    imaginary parts, as `~hypothesis.strategies.floats`.
+    """
+
+    check_valid_magnitude(min_magnitude, 'min_magnitude')
+    check_valid_magnitude(max_magnitude, 'max_magnitude')
+    check_valid_interval(
+        min_magnitude, max_magnitude,
+        'min_magnitude', 'max_magnitude'
+    )
+
+    if max_magnitude == float(u'inf'):
+        max_magnitude = None
+    if min_magnitude == 0:
+        min_magnitude = None
+
+    if allow_infinity is None:
+        allow_infinity = bool(max_magnitude is None)
+    elif allow_infinity and max_magnitude is not None:
+        raise InvalidArgument(
+            'Cannot have allow_infinity=%r with max_magnitude=%r' %
+            (allow_infinity, max_magnitude)
+        )
+
+    if allow_nan is None:
+        allow_nan = bool(min_magnitude is None and max_magnitude is None)
+    elif allow_nan and not (min_magnitude is None and max_magnitude is None):
+        raise InvalidArgument(
+            'Cannot have allow_nan=%r, min_magnitude=%r max_magnitude=%r' % (
+                allow_nan,
+                min_magnitude,
+                max_magnitude
+            )
+        )
+
+    if max_magnitude is None:
+        max_magnitude = float(u'inf')
+    if min_magnitude is None:
+        min_magnitude = 0
+
+    if allow_nan:
+        zi = draw(floats(allow_nan=True))
```

Comment:
I think this is incorrect if `allow_nan=True, allow_infinity=False`, and there's nothing banning that combination of arguments.

Reply:
added `allow_infinity=allow_infinity`